### PR TITLE
Use AddrSpacePtr to call WMMA intrinsics

### LINKED ISSUE #548: WMMA uses generic addressing
Not sure why I didn't notice this before...

```julia
using CUDAnative, CuArrays;

function kernel()
    shmem = @cuStaticSharedMem(Float32, (16, 16))
    conf = WMMA.Config{16, 16, 16, Float32}

    d_frag = WMMA.fill_c(Float32(0), conf)
    WMMA.store_d(pointer(shmem), d_frag, 16, WMMA.ColMajor, conf)

    return
end

CUDAnative.code_llvm(kernel, ())
CUDAnative.code_ptx(kernel, ())
```

gives:
```llvm
;  @ REPL[2]:2 within `kernel'
define void @julia_kernel_1() {
top:
;  @ REPL[2]:6 within `kernel'
; ┌ @ /home/tfaingna/.julia/dev/CUDAnative/src/device/cuda/wmma.jl:587 within `store_d'
; │┌ @ /home/tfaingna/.julia/dev/CUDAnative/src/device/cuda/wmma.jl:597 within `macro expansion'
; ││┌ @ /home/tfaingna/.julia/dev/CUDAnative/src/device/cuda/wmma.jl:158 within `llvm_wmma_store_d_col_m16n16k16_shared_stride_f32'
     call void @llvm.nvvm.wmma.m16n16k16.store.d.col.stride.f32.p3i8(i8* addrspacecast (i8 addrspace(3)* bitcast ([256 x float] addrspace(3)* @shmem__static_shmem_253 to i8 addrspace(3)*) to i8*), float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00, i32 16)
; └└└
;  @ REPL[2]:8 within `kernel'
  ret void
}
```

On the LLVM side, everything looks fine: the `p3` variant for shared memory is used.

```
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

	// .globl	julia_kernel_2  // -- Begin function julia_kernel_2
// shmem__static_shmem_253 has been demoted
                                        // @julia_kernel_2
.visible .func julia_kernel_2()
{
	.reg .f32 	%f<2>;
	.reg .b32 	%r<2>;
	.reg .b64 	%rd<3>;
	// demoted variable
	.shared .align 32 .b8 shmem__static_shmem_253[1024];
// %bb.0:                               // %top
	mov.u64 	%rd1, shmem__static_shmem_253;
	cvta.shared.u64 	%rd2, %rd1;
	mov.u32 	%r1, 16;
	mov.f32 	%f1, 0f00000000;
	wmma.store.d.sync.aligned.col.m16n16k16.f32 	[%rd2],{%f1, %f1, %f1, %f1, %f1, %f1, %f1, %f1}, %r1;
	ret;
                                        // -- End function
}
```

However, the NVPTX backend selected an `wmma.store.d.sync.aligned.col.m16n16k16.f32` and a `cvta.shared` instead of just using a `wmma.store.d.sync.aligned.col.m16n16k16.shared.f32`.
This seems to happen because of the addrspacecast to `i8*`, even though the intrinsic has argument type `i8 addrspace(3)*`. This confuses the NVPTX backend, which consequently selects the generic WMMA variant.

To fix this, I should either make sure that `ccall(..., llvmcall, ...)` respects addrspaces, or extend `Ref{T}` with an addrspace argument `Ref{T, A}`.

### PR DESCRIPTION
**Pull request by thomasfaingnaert**
Closes #548 
Depends on https://github.com/JuliaLang/julia/pull/34760

### CODE REVIEW

**Comment by maleadt** [comment]
aarch64 timed out, manually restarted.

File: src/device/cuda/wmma.jl [hunk 1]
```diff
@@ -103,7 +126,11 @@ for mat in ["a", "b", "c"],
 
     ccall_name = "extern $llvm_intr"
 
+    if addrspaceptr_available
+    @eval $func_name(src_addr, stride) = ccall($ccall_name, llvmcall, NTuple{$sz, $frag_ty}, (Core.AddrSpacePtr{$arr_ty, $addr_space_int}, Int32), convert(Core.AddrSpacePtr{$arr_ty, $addr_space_int}, src_addr), stride)
```
Comment (maleadt) [inline src/device/cuda/wmma.jl]: You can avoid these multiple definitions by implementing `cconvert` instead; that method is called already. Furthermore, using a variable for the destination type(Ref or AddrSpacePtr) you can get rid of the `if` too.

File: src/device/cuda/wmma.jl [hunk 2]
```diff
@@ -1,12 +1,16 @@
 export WMMA
 module WMMA
 
+import Base.convert
```
Comment (maleadt) [inline src/device/cuda/wmma.jl]: Just specify `function Base.convert` below instead.

**Comment by maleadt** [comment]
Looks great!
bors d+

**Comment by thomasfaingnaert** [comment]
bors try

**Comment by thomasfaingnaert** [comment]
Did a final check on v1.4 RC, everything still works correctly.
(Nightly is failing right now on base intrinsics, but you probably knew that already)
LGTM

bors merge

**Pull request merged**